src/map_to_curve.rs

```diff
@@ -6,31 +6,234 @@
 //! Defines the GroupWithMapToCurve trait to specify how to map a password to a
 //! curve point
 
+use crate::errors::InternalPakeError;
 use crate::group::Group;
+use crate::hash::Hash;
+use crate::serialization::i2osp;
 use curve25519_dalek::{edwards::EdwardsPoint, ristretto::RistrettoPoint};
-
+use digest::{BlockInput, Digest};
+use generic_array::typenum::Unsigned;
+use generic_array::GenericArray;
 use hkdf::Hkdf;
-use sha2::{Sha256, Sha512};
+use sha2::Sha256;
 
 /// A subtrait of Group specifying how to hash a password into a point
 pub trait GroupWithMapToCurve: Group {
-    /// transforms a password and optional pepper into a curve point
-    fn map_to_curve(password: &[u8], dst: Option<&[u8]>) -> Self;
-}
+    /// The ciphersuite identifier as dictated by
+    /// https://www.ietf.org/archive/id/draft-irtf-cfrg-voprf-05.txt
+    const SUITE_ID: usize;
 
-// TODO: incorporate expand_message_xmd from https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.txt
-// instead of using HKDF-extract here
+    /// transforms a password and domain separation tag (DST) into a curve point
+    fn map_to_curve(msg: &[u8], dst: &[u8]) -> Result<Self, InternalPakeError>;
+
+    /// Generates the contextString parameter as defined in
+    /// https://www.ietf.org/archive/id/draft-irtf-cfrg-voprf-05.txt
+    fn get_context_string(mode: u8) -> Vec<u8> {
+        [i2osp(mode as usize, 1), i2osp(Self::SUITE_ID, 2)].concat()
+    }
+}
 
 impl GroupWithMapToCurve for RistrettoPoint {
-    fn map_to_curve(password: &[u8], dst: Option<&[u8]>) -> Self {
-        let (hashed_input, _) = Hkdf::<Sha512>::extract(dst, password);
-        <Self as Group>::hash_to_curve(&hashed_input)
+    const SUITE_ID: usize = 0x01;
+
+    // Implements the hash_to_ristretto255() function from
+    // https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.txt
+    fn map_to_curve(msg: &[u8], dst: &[u8]) -> Result<Self, InternalPakeError> {
+        let uniform_bytes = expand_message_xmd::<Sha256>(msg, dst, 64)?;
+        Ok(<Self as Group>::hash_to_curve(
+            &GenericArray::clone_from_slice(&uniform_bytes[..]),
+        ))
     }
 }
 
 impl GroupWithMapToCurve for EdwardsPoint {
-    fn map_to_curve(password: &[u8], dst: Option<&[u8]>) -> Self {
-        let (hashed_input, _) = Hkdf::<Sha256>::extract(dst, password);
-        <Self as Group>::hash_to_curve(&hashed_input)
+    const SUITE_ID: usize = 0x09; // FIXME, seemingly unsupported by VOPRF RFC?
+    fn map_to_curve(msg: &[u8], dst: &[u8]) -> Result<Self, InternalPakeError> {
+        let (hashed_input, _) = Hkdf::<Sha256>::extract(Some(dst), msg);
+        Ok(<Self as Group>::hash_to_curve(&hashed_input))
+    }
+}
+
+// Computes ceil(x / y)
+fn div_ceil(x: usize, y: usize) -> usize {
+    let additive = match x % y != 0 {
+        true => 1,
+        false => 0,
+    };
```

Comment:
Would `if x % y == 0 { 0 } else { 1 }` be simpler? (or if we're really going for it, `(x%y != 0) as usize`)